processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -0,0 +1,713 @@
+import org.apache.druid.segment.index.semantic.Utf8ValueSetIndexes;
+import org.apache.druid.segment.index.semantic.ValueSetIndexes;
+import org.apache.druid.segment.vector.VectorColumnSelectorFactory;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedSet;
+
+public class TypedInFilter extends AbstractOptimizableDimFilter implements Filter
+{
+  private final String column;
+  private final ColumnType matchValueType;
+  @Nullable
+  private final List<?> unsortedValues;
+  private final Supplier<List<?>> lazyMatchValues;
+  @Nullable
+  private final Supplier<SortedSet<ByteBuffer>> lazyMatchValueBytes;
+  @Nullable
+  private final FilterTuning filterTuning;
+  private final Supplier<DruidPredicateFactory> predicateFactorySupplier;
+
+  @JsonIgnore
+  private final Supplier<byte[]> cacheKeySupplier;
+
+  /**
+   * Creates a new filter.
+   *
+   * @param column         column to search
+   * @param values         set of values to match. This collection may be reused to avoid copying a big collection.
+   *                       Therefore, callers should <b>not</b> modify the collection after it is passed to this
+   *                       constructor.
+   * @param matchValueType type of values contained in set
+   * @param filterTuning   optional tuning
+   */
+  @JsonCreator
+  public TypedInFilter(
+      @JsonProperty("column") String column,
+      @JsonProperty("matchValueType") ColumnType matchValueType,
+      @JsonProperty("values") @Nullable List<?> values,
+      @JsonProperty("sortedValues") @Nullable List<?> sortedValues,
+      @JsonProperty("filterTuning") @Nullable FilterTuning filterTuning
+  )
+  {
+    if (NullHandling.replaceWithDefault()) {
+      throw InvalidInput.exception(
+          "Invalid IN filter, typed in filter only supports SQL compatible null handling mode, set druid.generic.useDefaultValue=false to use this filter"
+      );
+    }
+    this.column = column;
+    if (column == null) {
+      throw InvalidInput.exception("Invalid IN filter, column cannot be null");
+    }
+    this.filterTuning = filterTuning;
+    this.matchValueType = matchValueType;
+    if (matchValueType == null) {
+      throw InvalidInput.exception("Invalid IN filter on column [%s], matchValueType cannot be null", column);
+    }
+    // one of sorted or not sorted
+    if (values == null && sortedValues == null) {
+      throw InvalidInput.exception(
+          "Invalid IN filter on column [%s], exactly one of values or sortedValues must be non-null",
+          column
+      );
+    }
+    if (sortedValues != null) {
+      this.unsortedValues = null;
+      this.lazyMatchValues = () -> sortedValues;
+    } else {
+      if (checkSorted(values, matchValueType)) {
```

Comment:
Do we need to check for duplicates too, or are duplicates ok in `lazyMatchValues`? Similar question for `sortValues`: does it need to dedupe or is just sorting ok?

Reply:
oops, `checkSorted` actually does mark it as not sorted if there are dupe values, but `sortValues` method used to dedupe and now it isn't, i should probably fix that, was lost when i switched to using `ObjectArrays.quickSort`.

that said, dupes wouldn't really break anything, just inefficient

Reply:
i went back to using a sorted set to order and dedupe since I could think of a very efficient way to do this otherwise other than creating my own modified version of quickSort that also dedupes

Reply:
Hmm, dedupe after quicksort is O(n); probably wouldn't notice it much next to the quicksort. It might even be faster to do quicksort + dedupe pass vs. doing tree-sort. Although, this probably doesn't matter _that_ much, since it's only happening once per query. The sorted set approach is IMO OK for this patch, esp. since that's the same thing the `in` filter is doing.

Reply:
yea, i toyed with a bit but didn't spend anytime measuring anything yet. I think the main thing that made me switch was the addition of the value coercion, with it too we iterate list once to make them all be the matchValueType, then the sort, then again to dedupe, so it seemed like maybe i should just use the set and coerce on insert and then turn that into a list.

Now i'm a bit curious though, so will do some measurements.

Reply:
ok, so got around to measuring it and went back to quicksort, even with coercion and dedupe is still better on bigger sets 

```
set:
Benchmark                           (filterSize)  Mode  Cnt     Score     Error  Units
InFilterBenchmark.sortFilterValues             1  avgt    3     0.021 ±   0.006  us/op
InFilterBenchmark.sortFilterValues            10  avgt    3     0.323 ±   0.035  us/op
InFilterBenchmark.sortFilterValues           100  avgt    3     4.521 ±   0.483  us/op
InFilterBenchmark.sortFilterValues          1000  avgt    3   104.984 ±  11.389  us/op
InFilterBenchmark.sortFilterValues         10000  avgt    3  1906.647 ± 922.912  us/op

quicksort:
Benchmark                           (filterSize)  Mode  Cnt     Score     Error  Units
InFilterBenchmark.sortFilterValues             1  avgt    3     0.021 ±   0.005  us/op
InFilterBenchmark.sortFilterValues            10  avgt    3     0.574 ±   0.083  us/op
InFilterBenchmark.sortFilterValues           100  avgt    3     4.678 ±   1.100  us/op
InFilterBenchmark.sortFilterValues          1000  avgt    3    84.306 ±  10.615  us/op
InFilterBenchmark.sortFilterValues         10000  avgt    3  1555.324 ± 228.895  us/op
```
(I didn't add this benchmark to the PR)